### PR TITLE
Use API GW as API configuration store

### LINKED ISSUE #1515: API GW : Variable name implies Cloudant-only support
In the backend routemgmt actions, access to the DB use function and variable names having "cloudant" in the label.  CouchDB should also work with these actions; so the label should be changed to signify broader DB support.

### LINKED ISSUE #1690: Refactor API GW actions to no longer use deprecated whisk SDK APIs
Replace whisk.error() with JavaScript promise reject().  May need to create a new promise.  Keep the same error context

Replace whisk.invoke() with the `openwhisk` NPM package's actions.invoke() method.
https://www.npmjs.com/package/openwhisk

### LINKED ISSUE #1697: Eliminate/Reduce redundant API GW configuration
API GW configuration is stored in both the API GW and in an OpenWhisk database.  Updating an existing configuration value across both stores is not currently atomic and is subject to being out of sync.

One possible solution is to implement an atomic operation across a chain of actions and API calls.  When an action in the chain fails, all prior actions should roll-back.

Another less complicated solution is to reduce or eliminate all of the redundant configuration so that the configuration operation is does not require a sequence of APIs/actions.

### PR DESCRIPTION
**Pull request by mdeuser**
Repackage routemgmt actions 
- As a .zip
 - Use common utilities module
Refactor routemgmt actions
 - No longer store API configuration in a separate database
 - Use the API GW as the API configuration data store
Small updates to automated tests

Fixes #1697
Fixes #1690 
Fixes #1515 

Final merge is dependent/blocked on https://github.com/openwhisk/apigateway/issues/56

### CODE REVIEW

**Comment by mdeuser** [comment]
@csantanapr PG2 895

**Comment by csantanapr** [comment]
@mdeuser as discussed offline:

We can't merge this change until apigateway api management interface implements the new filter functions being introduce in this PR.

Merging would mean any tests or users trying to create more than one api/basepath would be borken.

**Comment by mdeuser** [comment]
@csantanapr - the latest API GW docker image has the needed query filter support.  I've added a new test case to confirm API configuration when >1 APIs are defined.

PG3 189

**Comment by mdeuser** [comment]
PG3 190

**Comment by mdeuser** [comment]
@csantanapr - PG/Travis pass.  Ready for final review/merge

File: core/routemgmt/createApi/createApi.js [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  })
+  .then(function(swaggerApi) {
+    console.log('Final swagger API config: '+ JSON.stringify(swaggerApi));
+    return utils.addApiToGateway(gwInfo, tenantId, swaggerApi, gwApiId);
+  })
+  .then(function(gwApi) {
+    console.log('API GW configured with API');
+    var cliApi = utils.generateCliApiFromGwApi(gwApi).value;
+    console.log('createApi success');
+    return Promise.resolve(cliApi);
```
Comment (csantanapr) [inline core/routemgmt/createApi/createApi.js]: Need to check here what get's return to the CLI when creating apis using swagger vs. creating one endpoint. And be sure we have a test cases in place.

File: core/routemgmt/getApi/getApi.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  console.log('basepath/name : '+message.basepath);
+  console.log('relpath       : '+message.relpath);
+  console.log('operation     : '+message.operation);
+  console.log('outputFormat  : '+message.outputFormat);
+
+  // Issue a request to read API(s) from the API GW
+  // 1. Get the tenant ID associated with the specified namespace and optional tenant instance
+  // 2. Get the API(s) associated with the tenant ID and optional basepath/apiname
+  // 3. Format the API(s) per the outputFormat specification
+  return utils.getTenants(gwInfo, message.namespace, message.tenantInstance)
```
Comment (csantanapr) [inline core/routemgmt/getApi/getApi.js]: Need to set default value for `tenantInstance` to `openwhisk`

File: core/routemgmt/deleteApi/deleteApi.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  var tenantId;
+  return utils.getTenants(gwInfo, message.namespace, message.tenantInstance)
+  .then(function(tenants) {
+    // If a non-empty tenant array was returned, pick the first one from the list
+    if (tenants.length === 0) {
+      console.error('No Tenant found for namespace '+message.namespace);
+      return Promise.reject('No Tenant found for namespace '+message.namespace);
+    } else if (tenants.length > 1 ) {
+      console.error('Multiple tenants found for namespace '+message.namespace+' and tenant instance '+message.tenantInstance);
+      return Promise.reject('Internal error. Multiple API Gateway tenants found for namespace '+message.namespace+' and tenant instance '+message.tenantInstance);
```
Comment (csantanapr) [inline core/routemgmt/deleteApi/deleteApi.js]: Need to set default value for `tenantInstance` to `openwhisk`

File: core/routemgmt/common/utils.js [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+ * Returns:
+ *  array   : New array of API object - each in the specified format
+ */
+function transformApis(apis, format) {
+  var apisOutput;
+  try {
+    if (format.toLowerCase() === 'apigw') {
+      apisOutput = apis;
+    } else if (format.toLowerCase() === 'swagger') {
+      apisOutput = JSON.parse(JSON.stringify(apis));
```
Comment (csantanapr) [inline core/routemgmt/common/utils.js]: Might be able to use `Object.assign` instead and reap the performance benefits.

**Comment by mdeuser** [comment]
Travis OK.  PG3 205

File: core/routemgmt/createApi/createApi.js (hunk 1 above)
Comment (mdeuser) [inline core/routemgmt/createApi/createApi.js]: CLI is expecting to receive a single API description, not an array.  Test cases are in place to confirm successful API creation.  When retrieving the APIs (get/list) a results array is expected.

**Comment by mdeuser** [comment]
@csantanapr - critical comments are address
1. The `wsk api-experimental create` does indeed expect a single API definition, not an array.  Tests are in place
2. Code is updated to consistently use the tenant instance value; defaulting it to `openwhisk` for now.

**Comment by mdeuser** [comment]
PG3 205 - all API GW related tests passed.
@csantanapr

File: core/routemgmt/common/utils.js [hunk 5]
```diff
@@ -327,8 +328,8 @@ function getApis(gwInfo, tenantId, bpOrApiName) {
             } else {
               console.error('getApis: Invalid API GW response body; a JSON array was not returned');
               //FIXME MWD may happen when no apis are found and {} is returned:  https://github.com/openwhisk/apigateway/issues/59
-              //reject('Internal error. Invalid API Gateway response: Not an array');
-              resolve( [] );
+              //FIXME MWD reject('Internal error. Invalid API Gateway response: Not an array');
```
Comment (csantanapr) [inline core/routemgmt/common/utils.js]: We need to get with this guys to get this fix

**Comment by csantanapr** [comment]
squashed, rebase and merged fb65035a180c4846076c68a6a131cf81df17cc9b

**Pull request closed without merging**